### PR TITLE
Update MATLAB mapping with C++ proxy changes

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR updates MATLAB to use the latest C++ code - in particular the latest C++ code.

Ice for MATLAB stores proxies as void*. 

Prior to this PR, the void* is a pointer to a heap-allocated `shared_ptr<ObjectPrx>`, where the shared_ptr is never nullptr.
With this PR, the void* is a pointer to heap-allocated ObjectPrx.

### CODE REVIEW

File: matlab/src/Communicator.cpp [hunk 1]
```diff
@@ -59,15 +59,7 @@ Ice_Communicator_stringToProxy(void* self, const char* s, void** proxy)
...
-        auto p = deref<Ice::Communicator>(self)->stringToProxy(s);
-        if(p)
-        {
-            *proxy = createShared<Ice::ObjectPrx>(p);
-        }
-        else
-        {
-            *proxy = 0;
-        }
+        *proxy = createProxy(deref<Ice::Communicator>(self)->stringToProxy(s));
```
Comment (bernardnormier) [inline matlab/src/Communicator.cpp:62]: I added two overloads for createProxy:
 - the ObjectPrx overload never returns nullptr
 - the std::optional<ObjectPrx> overload returns a nullptr when the arg is nullopt

File: matlab/src/Util.h [hunk 2]
```diff
@@ -60,4 +60,32 @@ void* createShared(std::shared_ptr<T> p)
...
+{
+    return new Ice::ObjectPrx(std::move(p));
+}
+
+inline void* createProxy(std::optional<Ice::ObjectPrx> p)
+{
+    return p ? createProxy(std::move(p).value()) : nullptr;
+}
+
+inline Ice::ObjectPrx restoreProxy(void* p)
```
Comment (bernardnormier) [inline matlab/src/Util.h:73]: Maybe derefProxy / derefNullableProxy would be more consistent with the `deref<T>` function name?

Comment (externl) [reply matlab/src/Util.h:73]: I'm fine either way.

**Review by externl — APPROVED** [review]

File: matlab/src/Util.h [hunk 3]
```diff
@@ -60,4 +60,32 @@ void* createShared(std::shared_ptr<T> p)
     return new std::shared_ptr<T>(std::move(p));
 }
 
+inline void* createProxy(Ice::ObjectPrx p)
+{
+    return new Ice::ObjectPrx(std::move(p));
+}
+
+inline void* createProxy(std::optional<Ice::ObjectPrx> p)
+{
```
Comment (pepone) [inline matlab/src/Util.h:69]: Would be clear to pass as `const&` we never store p, or pass it to other methods

**Review by pepone — APPROVED** [review]

File: matlab/src/Util.h (hunk 3 above)
Comment (bernardnormier) [reply matlab/src/Util.h:69]: The implementation moves p, see below.

The general rule is:
 - if the implementation "keeps" the arg, pass it by value
 - if the implementation only reads the arg, pass it by const&

Comment (bernardnormier) [reply matlab/src/Util.h:69]: In particular, `.value()` below corresponds to the && overload:
https://en.cppreference.com/w/cpp/utility/optional/value (2)

Comment (pepone) [reply matlab/src/Util.h:69]: The implementation moves `p`, and calls `.value()` on the returned `rrefence`.

**Pull request merged**